Detectors/TPC/reconstruction/CMakeLists.txt

```diff
@@ -22,11 +23,25 @@ set(HEADERS
    include/${MODULE_NAME}/RawReader.h
    include/${MODULE_NAME}/SyncPatternMonitor.h
    include/${MODULE_NAME}/TrackTPC.h
+   include/${MODULE_NAME}/TPCCATracking.h
 )
 Set(LINKDEF src/TPCReconstructionLinkDef.h)
 Set(LIBRARY_NAME ${MODULE_NAME})
 set(BUCKET_NAME tpc_reconstruction_bucket)
 
+if(DEFINED O2_TPCCA_TRACKING_LIB_DIR)
+  include_directories(${O2_TPCCA_TRACKING_LIB_DIR}/include)
+  add_definitions(-DHAVE_O2_TPCCA_TRACKING_LIB -DHLTCA_STANDALONE -DHLTCA_TPC_GEOMETRY_O2)
+  FIND_LIBRARY(CATRACKING_LIB libO2TPCCATracking.so ${O2_TPCCA_TRACKING_LIB_DIR}/lib)
+  if(NOT CATRACKING_LIB)
+    message(FATAL_ERROR "Cannot find library libO2TPCCATracking.so")
+  endif()
+  link_libraries(${CATRACKING_LIB})
+  message("Building TPCCA tracking interface from standalone library in ${O2_TPCCA_TRACKING_LIB_DIR}")
+else()
+  message("Not building TPCCA tracking interface")
+endif()
```

Comment:
@davidrohr : Where is this dependency? Can we not just modify the o2.sh recipe to always build the tracking lib so that we avoid this conditional building? Otherwise, I agree with @matthiasrichter that testing of our code base will not be as efficient and we might not spot problems in the CI.

Reply:
@sawenzel @matthiasrichter : This is a bit complicated.
In total, there will have to be three pathces: one to AliRoot, one to O2, and one to the build recipes.

This pull request is only the O2 interface for the tracking, and at the moment it shall not be compiled, because it needs an AliRoot patch that is part of a larger AliRoot patchset from the HLT dev branch, which is still evaluated by the TPC.

The code for the tracking library is contained in AliRoot, and is also used by AliRoot.
However, the library itself does not have any AliRoot dependency.
One alternative would be to take all the tracking code out of AliRoot and place it in a separate repository used by AliRoot and O2, but @dberzano told me he would prefer not to add additional dependencies to AliRoot, therewore we decided to leave the code within AliRoot for the time being.

For building the library, there are two ways: It can be build standalone (without building AliRoot, one just needs to check out AliRoot to get the code), or it can also be built as part of the normal AliRoot cmake workflow. In both cases, it creates a library without any dependency on AliRoot.
Dario mentioned, they are thinking about adding AliRoot as optional dependency to O2, but this is not available yet. Therefore, I have for the time added the O2_TPCCA_TRACKING_LIB_DIR cmake flag. It can point to the AliRoot installation dir (when we use the optional AliRoot build, or it could also point to the standalone library installation dir).

For reference, I have modified the o2 build recipe here: https://github.com/davidrohr/alidist/commit/e64a167fadd44c5f223387d83a51524fa2aedde8
This is a prototype that makes AliRoot a dependency for O2, and sets the O2 cmake setting to build against that library. It still needs to be discussed with Dario how the recipe will have to look in the end.
The AliRoot version with all patches is currently here: https://github.com/davidrohr/AliRoot/tree/dev

The O2 part is basically done, therefore I have opened this pull request now, such that people who want to play with the tracking can use at least the official O2 and only need to use my AliRoot branch, and also to simplify maintenance of the code.

Reply:
@davidrohr : Thanks for the explanation; I think there is a strong user story in software engineering which is "being able to build the whole software stack in one go and being able to actually run the code without complications".

So if you say that the hlt code could be build standalone from the AliRoot sources, I would suggest to make a recipe that does precisely this and make this  a requirement for O2 (at least with the default-o2). Or we make AliRoot a dependency.

Reply:
Hi @sawenzel ,
I totally agree, and that is one of the reasons (besides faster debugging), that I created the standalone build.

One could for instance use the AliRoot build when the future optional AliRoot dependency is enabled, and use the standalone build otherwise. That would allow us to get rid of the conditional compilation. However, some people didn't like that you would have to check out the large AliRoot repo in both cases.

But for now, since AliRoot master is insuffiecient for building this, I think we have to live with conditional compilation for the time being. In this way, we do not disturb normal users, but we enable those who need the tracking to build it. Once we have the large patch in AliRoot and a proper O2 build recipe, I would definitely remove the conditional compilation.

Cheers
David

Reply:
Hi David, 

Sorry if that's a dumb question, but I assume  this AliRoot dependency is anyway (whatever the form it will take) a "short/medium-term" measure only,  right ?

Reply:
Hi Laurent,

I cannot really tell.
From our side, I would say ASAP, and the tracking part is definitely ready.
But this TPC validation is ongoing since quite some time (for reference, here is the JIRA: https://alice.its.cern.ch/jira/browse/ATO-398), and it required some fixes already. Everything we know about is fixed, so I hope this can be merged soon, but I do not feel qualified to give a reasonable time estimate.